// Copyright 2024 The Ray Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

option cc_enable_arenas = true;

package ray.rpc;

import "google/protobuf/struct.proto";

// Metadata for a Ray Train run, including its details, status, and run settings
message ExportTrainRunEventData {
  // Version of the message schema
  uint32 schema_version = 1;

  // Enumeration of the possible statuses for a Train run
  enum RunStatus {
    // Default value required by proto3
    RUN_STATUS_UNSPECIFIED = 0;

    // Active States
    INITIALIZING = 1;
    SCHEDULING = 2;
    RUNNING = 3;
    RESTARTING = 4;
    RESIZING = 5;

    // Terminal States
    FINISHED = 6;
    ERRORED = 7;
    ABORTED = 8;
  }

  // Unique identifier for the Train run
  string id = 2;
  // Human-readable name assigned to the Train run
  string name = 3;
  // The Ray Job ID associated with this Train run
  bytes job_id = 4;
  // Unique ID of the actor managing the Train run
  bytes controller_actor_id = 5;
  // The current execution status of the Train run
  RunStatus status = 6;
  // Additional details about the current status, including error messages if applicable
  optional string status_detail = 7;
  // The timestamp when the Train run started in ns.
  uint64 start_time_ns = 8;
  // The timestamp when the Train run ended in ns. If not set, the run is still
  // in progress
  optional uint64 end_time_ns = 9;
  // The path to the log file for the Train run controller.
  optional string controller_log_file_path = 10;
  // Version of Ray Train
  uint32 ray_train_version = 11;

  // Dashboard panel configuration for visualizing Train run data
  message DashboardPanelMetadata {
    // Unique identifier for the panel
    string id = 1;
    // Display name of the panel
    string title = 2;
  }

  // List of TrainRun dashboard panels
  repeated DashboardPanelMetadata train_run_panels = 12;
  // List of dashboard panels to include for TrainWorkers
  repeated DashboardPanelMetadata train_worker_panels = 13;

  // Framework versions
  map<string, string> framework_versions = 14;

  // Run settings
  RunSettings run_settings = 15;

  // Settings for a Train run, primarily consisting of configs set before a train run starts.
  message RunSettings {
    // The user defined train loop config for a Train run.
    // NOTE: This is expected to represent a JSON-serializable dict from Python.
    optional string train_loop_config = 1;
    // The backend config for a Train run, varies with the framework (e.g. TorchConfig)
    BackendConfig backend_config = 2;
    // The scaling config for this Train run.
    ScalingConfig scaling_config = 3;
    // A list of dataset names for a Train run.
    repeated string datasets = 4;
    // The data config for a Train run.
    DataConfig data_config = 5;
    // Run configuration for this Train run.
    RunConfig run_config = 6;
  }

  // Backend config for a Train run, varies with the framework (e.g. TorchConfig).
  message BackendConfig {
    // Training Framework used for the train run
    enum TrainingFramework {
      // Default framework
      TRAINING_FRAMEWORK_UNSPECIFIED = 0;

      TORCH = 1;
      JAX = 2;
      TENSORFLOW = 3;
      XGBOOST = 4;
      LIGHTGBM = 5;
    }

    // The training framework for this backend config.
    TrainingFramework framework = 1;
    // Training framework-specific configuration fields.
    google.protobuf.Struct config = 2;
  }

  // Scaling config for a Train run.
  message ScalingConfig {
    message StringMap {
      map<string, string> values = 1;
    }
    message StringFloatMap {
      map<string, double> values = 1;
    }

    // The number of workers for the Train run.
    uint64 num_workers = 1;
    // Whether to use GPUs for the Train run.
    bool use_gpu = 2;
    // The resources per worker for a Train run.
    StringFloatMap resources_per_worker = 3;
    // The placement strategy for the Train run.
    string placement_strategy = 4;
    // The accelerator type for the Train run.
    optional string accelerator_type = 5;
    // Whether to use TPUs for the Train run.
    bool use_tpu = 6;
    // The topology for the Train run.
    optional string topology= 7;
    // The bundle label selector for the Train run.
    repeated StringMap bundle_label_selector = 8;
  }

  // Configuration for dataset splitting and execution options within Ray Train.
  message DataConfig {
    // Marker for selecting all datasets.
    message All {}
    message StringList {
      repeated string values = 1;
    }
    oneof datasets_to_split {
      // Split all datasets.
      All all = 1;
      // Split only the listed datasets (by dataset name).
      StringList datasets = 2;
    }
    // Data execution options.
    google.protobuf.Struct execution_options = 3;
    // Whether to enable shard locality optimization.
    bool enable_shard_locality = 4;
  }

  // Run configuration parameters for a Train run
  message RunConfig {
    // The failure config for a Train run.
    FailureConfig failure_config = 1;
    // The worker runtime env for a Train run.
    google.protobuf.Struct worker_runtime_env = 2;
    // The checkpoint config for a Train run.
    CheckpointConfig checkpoint_config = 3;
    // The storage path for a Train run.
    string storage_path = 4;
  }

  // Failure config for a Train run.
  message FailureConfig {
    // The maximum number of failures for a Train run.
    int32 max_failures = 1;
    // The maximum number of controller failures to tolerate.
    int32 controller_failure_limit = 2;
  }

  // Checkpoint config for a Train run.
  message CheckpointConfig {
    enum CheckpointScoreOrder {
      // Default value required by proto3
      SCORE_ORDER_UNSPECIFIED = 0;

      MAX = 1;
      MIN = 2;
    }

    // The number of most recent checkpoints to keep. Older checkpoints may be deleted.
    optional uint64 num_to_keep = 1;
    // Attribute used to score and rank checkpoints; can be a metric key or attribute.
    optional string checkpoint_score_attribute = 2;

    // Either “max” or “min”. If “max”/”min”, then checkpoints with highest/lowest values
    // of the checkpoint_score_attribute will be kept.
    CheckpointScoreOrder checkpoint_score_order = 3;
  }

}

// Metadata for an individual attempt to execute a Train run
message ExportTrainRunAttemptEventData {
  // Version of the message schema
  uint32 schema_version = 1;

  // Enumeration of the possible statuses for a Train run attempt
  enum RunAttemptStatus {
    // Default value required by proto3
    RUN_ATTEMPT_STATUS_UNSPECIFIED = 0;

    // Active States
    PENDING = 1;
    RUNNING = 2;

    // Terminal States
    FINISHED = 3;
    ERRORED = 4;
    ABORTED = 5;
  }

  // Enumeration of the statuses for a Train worker actor
  enum ActorStatus {
    // Default value required by proto3
    ACTOR_STATUS_UNSPECIFIED = 0;

    // The actor is currently active
    ALIVE = 1;
    // The actor is no longer active
    DEAD = 2;
  }

  // Resources allocated for a Train worker or run
  message TrainResources {
    // A dictionary specifying the types and amounts of resources allocated
    // (e.g., CPU, GPU)
    map<string, double> resources = 1;
  }

  // Metadata about a Ray Train worker
  message TrainWorker {
    // The global rank of the worker in the training cluster
    int32 world_rank = 1;
    // The local rank of the worker on its node
    int32 local_rank = 2;
    // The rank of the worker's node in the cluster
    int32 node_rank = 3;
    // The unique ID of the worker's actor
    bytes actor_id = 4;
    // The unique ID of the node where the worker is running
    bytes node_id = 5;
    // The IP address of the node where the worker is running
    string node_ip = 6;
    // The process ID of the worker
    int32 pid = 7;
    // A list of GPU IDs allocated to the worker
    repeated int32 gpu_ids = 8;
    // The current status of the worker actor
    optional ActorStatus status = 9;
    // The resources allocated to this Train worker
    TrainResources resources = 10;
    // The path to the log file for the Train worker
    optional string log_file_path = 11;
  }

  // Unique identifier for the parent Train run
  string run_id = 2;
  // Unique identifier for this specific Train run attempt
  string attempt_id = 3;
  // The current execution status of the Train run attempt
  RunAttemptStatus status = 4;
  // Additional details about the status, including error messages if applicable
  optional string status_detail = 5;
  // The timestamp when the Train run attempt started in ns.
  uint64 start_time_ns = 6;
  // The timestamp when the Train run attempt ended in ns. If null, the attempt
  // is still ongoing
  optional uint64 end_time_ns = 7;
  // The resources (e.g., CPU, GPU) allocated to the Train run attempt
  repeated TrainResources resources = 8;
  // List of Train workers participating in this attempt, sorted by global ranks
  repeated TrainWorker workers = 9;
  // Version of Ray Train
  uint32 ray_train_version = 10;
}
